controller/kv/helper.go

```diff
@@ -1582,6 +1602,133 @@ func (m clusterHelper) GetScannerDB(store string) []*share.CLUSScannerDB {
+		for _, key := range keys {
+			var scanner share.CLUSScanner
+			value, _, err := m.get(key)
+			if err != nil {
+				log.WithFields(log.Fields{"key": key, "error": err}).Error("Failed to get scanner")
+				continue
+			}
+			if value != nil {
+				if err := nvJsonUnmarshal(key, value, &scanner); err != nil {
+					log.WithFields(log.Fields{"key": key, "error": err}).Error("Failed to unmarshal scanner")
+					continue
+				}
+				if scanner.ID != share.CLUSScannerDBVersionID {
+					scanners = append(scanners, &scanner)
+				}
+			}
+		}
+	}
+
+	return scanners
+}
+
+func (m clusterHelper) PickLeastLoadedScanner(ScannerHealthCheckFunc func(string, time.Duration) error) (*share.CLUSScanner, error) {
+	scanners := m.GetAvailableScanners()
+	if len(scanners) == 0 {
+		return nil, share.ErrNoScannerFound
+	}
+
+	maxScanCredits := -1
+	var selectedScanner *share.CLUSScanner
+
+	for _, scanner := range scanners {
+		if scanner == nil {
+			log.Error("scanner is nil, skipping")
+			continue
+		}
+		scanCredits := scanner.ScanCredit
+
+		if ScannerHealthCheckFunc != nil {
+			if err := ScannerHealthCheckFunc(scanner.ID, time.Second*10); err != nil {
+				log.WithFields(log.Fields{"scanner": scanner.ID, "error": err}).Error("Failed to ping scanner")
+				continue
+			}
+		}
+
+		if scanCredits > maxScanCredits && scanCredits > 0 {
+			maxScanCredits = scanCredits
+			selectedScanner = scanner
+		}
+	}
+
+	if selectedScanner == nil {
+		return nil, share.ErrNoScannerAvailable
+	}
+
+	selectedScanner, rev, err := m.GetScannerRev(selectedScanner.ID)
+	if err != nil {
+		log.WithFields(log.Fields{"error": err, "scanner": selectedScanner.ID}).Error("Failed to get scanner")
+		return nil, err
+	}
+
+	if selectedScanner.ScanCredit <= 0 {
+		return nil, fmt.Errorf("scanner %s has no available credit", selectedScanner.ID)
+	}
+
+	// Decrement the scan credit
+	selectedScanner.ScanCredit--
+
+	// Update the scanner in KV store
+	key := share.CLUSScannerKey(selectedScanner.ID)
+	value, err := json.Marshal(selectedScanner)
+	if err != nil {
+		return nil, err
+	}
+	if err := cluster.PutRev(key, value, rev); err != nil {
+		log.WithFields(log.Fields{"error": err, "scanner": selectedScanner.ID}).Error("Failed to update scanner credit")
+		return nil, err
+	}
+
+	return selectedScanner, nil
+}
+
+func (m clusterHelper) ReleaseScanCredit(scannerId string) error {
+	// Infinite retry loop to ensure credit is never lost
+	// This is critical because losing credits would cause scanner starvation
+	key := share.CLUSScannerKey(scannerId)
+
+	for {
+		// Get current scanner state with revision
+		value, rev, err := m.get(key)
+		if err != nil || value == nil {
+			// Scanner has been deleted - credit is implicitly released with the scanner
+			// This is the only successful exit condition besides successful Put
+			log.WithFields(log.Fields{"scanner": scannerId}).Debug("Scanner not found during credit release, likely deleted")
```

Comment:
I think you have to check the error before claiming the scanner is not found.  The `err` might indicate it's a network issue or something like that.  For those kind of issues you probably want to retry where possible.